File: src/main/java/io/serieznyi/intellij/factorioapicompletion/intellij/FactorioStartupActivity.java
package io.serieznyi.intellij.factorioapicompletion.intellij;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.startup.StartupActivity;
import io.serieznyi.intellij.factorioapicompletion.core.PrototypesService;
import io.serieznyi.intellij.factorioapicompletion.intellij.service.ApiService;
import io.serieznyi.intellij.factorioapicompletion.intellij.service.FactorioDataService;
import io.serieznyi.intellij.factorioapicompletion.intellij.service.PrototypeService;
import org.jetbrains.annotations.NotNull;

public class FactorioStartupActivity implements StartupActivity {
    @Override
    public void runActivity(@NotNull Project project) {
        FactorioState config = FactorioState.getInstance(project);

        if (config.integrationActive) {
            boolean update = FactorioDataService.getInstance(project).checkForUpdate();
            ApiService.getInstance(project).checkForUpdate();

            if (update) {
                PrototypeService.getInstance(project).removeLibraryFiles();
                PrototypeService.getInstance(project).checkForUpdate();
            }

            // reload core/base prototypes
            PrototypesService.getInstance(project).reloadIndex();
        }
    }
}

File: src/main/kotlin/io/serieznyi/intellij/factorioapicompletion/intellij/FactorioStartupActivity.kt
package io.serieznyi.intellij.factorioapicompletion.intellij

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.EDT
import com.intellij.openapi.application.WriteAction
import com.intellij.openapi.application.runInEdt
import com.intellij.openapi.components.service
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.ProjectActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class FactorioStartupActivity: ProjectActivity {
    override suspend fun execute(project: Project) {
//        ProgressManager.getInstance().run { project.service<ApiService>().downloadApi() }
//
//        runInEdt {
//            project.service<ApiService>().downloadApi()
//        }
//
//        WriteAction.run<RuntimeException> {
//            ApplicationManager.getApplication().runWriteAction {
//
//            }
//        }
    }
}